### PR TITLE
[SPARK-35368][SQL] Update histogram statistics for RANGE operator for stats estimation

### PR DESCRIPTION
**Pull request by shahidki31**
### What changes were proposed in this pull request?
Update histogram statistics for RANGE operator stats estimation
### Why are the changes needed?
If histogram optimization is enabled, this statistics can be used in various cost based optimizations.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UTs. Manual test.

### CODE REVIEW

**Comment by shahidki31** [comment]
cc @wangyum @cloud-fan @HyukjinKwon  Kindly review

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -759,6 +759,10 @@ case class Range(
   }
 
   override def computeStats(): Statistics = {
+    if (!conf.cboEnabled) {
+      return Statistics(sizeInBytes = LongType.defaultSize * numElements)
+    }
+
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:765]: We need this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 2]
```diff
@@ -767,14 +771,44 @@ case class Range(
       } else {
         (start + (numElements - 1) * step, start)
       }
-      val colStat = ColumnStat(
+      var colStat = ColumnStat(
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:774]: Please avoid using `var` where possible.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 3]
```diff
@@ -767,14 +771,44 @@ case class Range(
...
         distinctCount = Some(numElements),
         max = Some(maxVal),
         min = Some(minVal),
         nullCount = Some(0),
         avgLen = Some(LongType.defaultSize),
         maxLen = Some(LongType.defaultSize))
 
+      if (conf.histogramEnabled) {
+
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:783]: please remove this blank.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 4]
```diff
@@ -767,14 +771,44 @@ case class Range(
...
+        def getRangeValue(index: Int): Long = {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:784]: How about extracting this inner method as a private method?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 1 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: I removed the condition as we have all the statistics available for the operator. I have added the condition earlier because, if cbo is not enabled, other operators are not returning other statistics. Do we also need to remove histogram check as well?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 2 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 3 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Removed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 4 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 1 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: I removed histogram config check as well. Because histogram information is readily available. Also refactored the code and Its

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 5]
```diff
@@ -774,13 +774,17 @@ case class Range(
       } else {
         (start + (numElements - 1) * step, start)
       }
+
+      val histogram = getHistogramStatistics
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:778]: Let's check `conf.histogramEnabled` here;
```
val histogram = if (conf.histogramEnabled) {
  Some(computeHistogramStatistics())
} else {
  None
}
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 6]
```diff
@@ -789,6 +793,36 @@ case class Range(
...
+    val percentileArray = (0 to numBins).map(i => i * height).toArray
+
+    var binId = 0
+    val binArray = new Array[HistogramBin](numBins)
+    var lowerIndex = percentileArray(0)
+    var lowerBinValue = getRangeValue(0)
+    while (binId < numBins) {
+      val upperIndex = percentileArray(binId + 1)
+      val upperBinValue = getRangeValue(math.max(math.ceil(upperIndex).toInt - 1, 0))
+      val ndv = math.max(math.ceil(upperIndex).toInt - math.ceil(lowerIndex).toInt, 1)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:808]: Please do not repeat `math.ceil(upperIndex)` here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 5 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:778]: I feel its better to wrap a return value of `computeHistogramStatistics ` with `Some` in the caller side because the method always return a value.
https://github.com/apache/spark/pull/32498/files#diff-0ac2c89f8cc0d00e8fe7717b01697f36f20fe8abf2def09bcfde0ad84b30e467R814

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 7]
```diff
@@ -789,6 +793,36 @@ case class Range(
     }
   }
 
+  private def getHistogramStatistics = {
+    val numBins = conf.histogramNumBins
+    val height = numElements.toDouble / numBins
+    val percentileArray = (0 to numBins).map(i => i * height).toArray
+
+    var binId = 0
```
[the remark this answers is not in the document]
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Yes, I updated this with `zipWithIndex`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 8]
```diff
@@ -789,6 +793,36 @@ case class Range(
     }
   }
 
+  private def getHistogramStatistics = {
```
[the remark this answers is not in the document]
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 5 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 6 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Yes, updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 9]
```diff
@@ -789,6 +797,38 @@ case class Range(
...
   }
 
+  private def computeHistogramStatistics() = {
+    val numBins = conf.histogramNumBins
+    val height = numElements.toDouble / numBins
+    val percentileArray = (0 to numBins).map(i => i * height).toArray
+
+    val binArray = new Array[HistogramBin](numBins)
+    var lowerIndex = percentileArray.head
+    var lowerBinValue = getRangeValue(0)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:807]: It looks we can remove `var` by using `foldLeft` instead of `foreach`. Could you?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 10]
```diff
@@ -789,6 +797,38 @@ case class Range(
...
+      binArray(binId) = HistogramBin(lowerBinValue, upperBinValue, ndv)
+
+      lowerBinValue = upperBinValue
+      lowerIndex = upperIndex
+    }
+    Histogram(height, binArray)
+  }
+
+  // Utility method to compute histogram
+  private def getRangeValue(index: Int): Long = {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:824]:
```
  private def getRangeValue(index: Int): Long = {
    assert(index >= 0, "index must be greater than and equal to 0")
    if (step == 0) {
      start + (numElements.toLong - index - 1) * step
    } else {
      start + index * step
    }
  }
```
?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala [hunk 11]
```diff
@@ -97,12 +121,24 @@ class BasicStatsEstimationSuite extends PlanTest with StatsEstimationTestBase {
...
+                avgLen = Some(LongType.defaultSize),
+                histogram = histogram)))))
+    val extraConfig = Map(SQLConf.HISTOGRAM_ENABLED.key -> "true",
+      SQLConf.HISTOGRAM_NUM_BINS.key -> "3")
+    checkStats(range, expectedStatsCboOn = rangeStats,
+      expectedStatsCboOff = rangeStats, extraConfig)
   }
 
   test("range with negative step where end minus start not divisible by step") {
+
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala:133]: nit: please revert this change.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala [hunk 12]
```diff
@@ -77,12 +92,21 @@ class BasicStatsEstimationSuite extends PlanTest with StatsEstimationTestBase {
...
+    checkStats(range, expectedStatsCboOn = rangeStats,
+      expectedStatsCboOff = rangeStats, extraConfig)
   }
 
   test("range with negative step") {
     val range = Range(-10, -20, -2, None)
+    val histogramBins = new Array[HistogramBin](3)
+    histogramBins(0) = HistogramBin(-18.0, -16.0, 2)
+    histogramBins(1) = HistogramBin(-16.0, -12.0, 2)
+    histogramBins(2) = HistogramBin(-12.0, -10.0, 1)
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala:108]: nit:
```
    val histogramBins = Array(
      HistogramBin(-18.0, -16.0, 2),
      HistogramBin(-16.0, -12.0, 2),
      HistogramBin(-12.0, -10.0, 1))
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala [hunk 13]
```diff
@@ -97,12 +121,24 @@ class BasicStatsEstimationSuite extends PlanTest with StatsEstimationTestBase {
...
   }
 
   test("range with negative step where end minus start not divisible by step") {
+
     val range = Range(-10, -20, -3, None)
+
+    val histogramBins = new Array[HistogramBin](3)
+    histogramBins(0) = HistogramBin(-19.0, -16.0, 2)
+    histogramBins(1) = HistogramBin(-16.0, -13.0, 1)
+    histogramBins(2) = HistogramBin(-13.0, -10.0, 1)
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala:139]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala [hunk 14]
```diff
@@ -283,14 +326,17 @@ class BasicStatsEstimationSuite extends PlanTest with StatsEstimationTestBase {
   private def checkStats(
       plan: LogicalPlan,
       expectedStatsCboOn: Statistics,
-      expectedStatsCboOff: Statistics): Unit = {
-    withSQLConf(SQLConf.CBO_ENABLED.key -> "true") {
+      expectedStatsCboOff: Statistics,
+      extraConfigs: Map[String, String] = Map.empty): Unit = {
+
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala:331]: nit: remove this unnecessary change

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala (hunk 12 above)
Comment (maropu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala:108]: Could you check if `range.numElements` and the total `ndv` are the same?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 9 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Yes, updated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 10 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:825]: Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala (hunk 11 above)
Comment (shahidki31) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala]: Done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala (hunk 12 above)
Comment (shahidki31) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala]: Added assert to check if `range.numElements` and `ndv` are same

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala (hunk 13 above)
Comment (shahidki31) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala]: Updated

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala (hunk 14 above)
Comment (shahidki31) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala:337]: I am not sure I understand you here. Do we need to directly put the histogram configs inside this method? By default histogram is disabled and number of bins default value is 254.

Comment (shahidki31) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/statsEstimation/BasicStatsEstimationSuite.scala]: Yes, removed the extra line

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 15]
```diff
@@ -789,6 +797,40 @@ case class Range(
...
+          val upperIndexPos = math.ceil(upperIndex).toInt - 1
+          val lowerIndexPos = math.ceil(lowerIndex).toInt - 1
+
+          val upperBinValue = getRangeValue(math.max(upperIndexPos, 0))
+          val ndv = math.max(upperIndexPos - lowerIndexPos, 1)
+          binArray(binId) = HistogramBin(lowerBinValue, upperBinValue, ndv)
+          // Update the lowerIndex and lowerBinValue with upper ones for the next iteration.
+          (upperIndex, upperBinValue)
+      }
+    Histogram(height, binArray)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:821]: nit: how about rewriting it like this?
```

    val lowerIndexInitial = percentileArray.head
    val lowerBinValueInitial = getRangeValue(0)
    val (_, _, binArray) = percentileArray.tail
      .foldLeft((lowerIndexInitial, lowerBinValueInitial, Seq.empty[HistogramBin])) {
      case ((lowerIndex, lowerBinValue, binAr), upperIndex) =>
        // Integer index for upper and lower values in the bin.
        val upperIndexPos = math.ceil(upperIndex).toInt - 1
        val lowerIndexPos = math.ceil(lowerIndex).toInt - 1

        val upperBinValue = getRangeValue(math.max(upperIndexPos, 0))
        val ndv = math.max(upperIndexPos - lowerIndexPos, 1)
        // Update the lowerIndex and lowerBinValue with upper ones for the next iteration.
        (upperIndex, upperBinValue, binAr :+ HistogramBin(lowerBinValue, upperBinValue, ndv))
    }
    Histogram(height, binArray.toArray)
```

**Comment by maropu** [comment]
Looks fine otherwise.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 16]
```diff
@@ -789,6 +797,40 @@ case class Range(
...
+  }
+
+  // Utility method to compute histogram
+  private def getRangeValue(index: Int): Long = {
+    assert(index >= 0, "index must be greater than and equal to 0")
+    if (step <= 0) {
+      start + (numElements.toLong - index - 1) * step
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:828]: I'm probably missing something, but isn't this more like:
```
(start + numElements.toLong - 1) + index * step
```
This tries to step backwards from the end, right?
Also I assume step = 0 is an error actually, maybe it's caught earlier, but the result should be `start` always in that case anyway, so `step < 0` as a condition?

Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:827]: >This tries to step backwards from the end, right? step is negative

Yes, we reverse the range values if the step is negative to compute histogram statistics.

>I'm probably missing something, but isn't this more like:
    start + (numElements.toLong - 1) * (-step) + index * step

I am not sure this gives the right range values if the step is negative. 
For eg: start = 1 , step = -2, and numElements = 4, Range values  = [1, -1, -3, -5]
But for computing histogram we need values like this [-5, -3, -1, 1]. So if the index = 0, it should return -5.

With the formula in the PR => 1 + (4 - 0 -1 )* -2= -5
 With the formula you have given above  => 1 + (4 -1) *-(-2) + 0 = 7

So I think the formula in the PR seems correct?

Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:827]: >Also I assume step = 0 is an error actually, maybe it's caught earlier, but the result should be start always in that case anyway, so step < 0 as a condition?

Yes, if step=0 both the conditions will return same start element. I have updated the condition to step < 0.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 15 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: Updated. Thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 16 above)
Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:827]: OK, I thought it might mean the desired result was [7, 5, 3, 1] in this case. You step back from the start. But it has to be reversed? OK if it needs to be [-5, -3, -1, 1] yeah that looks right. A comment to this effect might be helpful.

Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:827]: Yes. Sure I will add a comment.

**Review by maropu — APPROVED** [review]
cc: @srowen

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 17]
```diff
@@ -789,6 +797,39 @@ case class Range(
     }
   }
 
+  private def computeHistogramStatistics() = {
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:800]: Trivial: you could put a return type here.
Is there any concern with the overhead of computing the histogram? It only happens when enabled of course. I just wonder if we need to micro-optimize the loop below with something more performant if this is critical, but maybe it isn't.

**Review by srowen — COMMENTED** [review]
Seems fine to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 17 above)
Comment (shahidki31) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: > Trivial: you could put a return type here.Yes, I added the return type

Yes, I updated with the return type.

> Is there any concern with the overhead of computing the histogram? It only happens when enabled of course. I just wonder if we need to micro-optimize the loop below with something more performant if this is critical, but maybe it isn't.

This is not a performance intensive path. This will execute during planning phase. Also the loop size is numBins, which is 254 by default. So I am not sure we need to micro-optimise the loop?

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
Thank you, @shahidki31 . Merged to master.

**Comment by shahidki31** [comment]
Thanks a lot @maropu , @srowen  for the reviews.
